docs/swagger/common/schemas/circuit.json

```diff
@@ -0,0 +1,59 @@
+{
+	"type": "object",
+	"properties": {
+		"gates": {
+			"type": "array",
+			"items": {
+				"type": "string",
+				"description": "Gates in the circuit"
+			}
+		},
+		"circuit": {
+			"type": "array",
+			"items": {
+				"type": "object",
+				"properties": {
+					"type": {
+						"type": "string",
+						"description": "Type of the gate in the circuit",
+						"example": "x"
+					},
+					"time": {
+						"type": "integer",
+						"example": 3
+					},
+					"targets": {
+						"type": "array",
+						"items": {
+							"type": "integer",
+							"example": 1
+						}
+					},
+					"controls": {
+						"type": "array",
+						"items": {
+							"type": "string"
+						}
+					}
+				}
+			}
+		},
+		"qubits": {
+			"type": "integer",
+			"description": "Number of qubits in the circuit",
+			"example": 2
+		},
+		"input": {
+			"type": "array",
+			"description": "The binary state of each qubit input",
+			"items": {
+				"type": "integer"
+			}
+		},
+		"version": {
+			"type": "integer",
+			"description": "Version of the circuit schema",
+			"example": 1
+		}
+	}
```

Comment:
Can we make it so that all API returns a _status_ property which can be checked upon, simillar to how you've been doing it in most of the API. It makes it easier and is much more kind to have a reliable way to say if an error occurred or not. Meaning, you can always check the _status_ property to see if there is anything wrong. If there isn't anything wrong, proceed to do whatever based on what you know you're supposed to get.

Another thing, it would be nice to put anything that you are expected to receive under _data_ property so it's more consistent where exactly the actually data that you care about is, again makes it easier for people. 

Really, [Jsend](https://github.com/omniti-labs/jsend) is a nice way to do this but we don't have to do it exactly. I just think the concept of one property always for data and one property always an indicator if there is an error is very easy and good. 

Also, bonus for having a place for error messages (You're already doing that I think). Maybe it would be nice for making those messages user friendly so that the frontend can simply display the message without any weird translation about backend messages, without trying to understand what to display back considering some error and without doing double work. It also might be possible to have an _error_ property and an _userFriendlyError_ property.

Reply:
Non of the endpoints are duplicates. We are using tags in the swagger file to collect the endpoints into sections and some endpoints have more than one tag depending on functionality so can be in 2 sections.

Generally status codes should be the way to check the response status but we can definitely  try and normalise API responses into the below data structure.

```json
{
  "status": "OK",
  "data": {
    "some": "data",
    "for": "endpoint"
  }
}
```

For the error API responses for debugging purposes we should give the full error string. I think we can add an error code to the response that is more understandable to at least easily check what the error was without having to parse the error string.

```json
{
  "status": "ERROR",
  "errorCode": "user-exists",
  "error": "the original error string that triggered the response"
}
```

But i still think that error messaging in the frontend should be controlled by the frontend as the error api response is an api response and not a frontend error string. 

Just a note on error responses, i left allot of them quite basic since alot of error scenarios will be discovered while developing the endpoints.

Reply:
Also on the point of removing `/circuit/{circuitId}/update` and using more of the HTTP verbs under a single endpoint i prefer to use a pattern of the endpoint being more descriptive of the action it is taking rather than always needing to check the specification. 

Also on the note of using POST rather than PUT. I would rather use POST since there will always be some kind of data processing. PUT as a HTTP verb is for stateless transactions and i cant really think of a scenario where we would ever want to do a stateless transaction with the API.